### PR TITLE
Feature/promo codes tags

### PR DESCRIPTION
**Pull request by romanetar**
ref https://tipit.avaza.com/project/view/203215#!tab=task-pane&task=3082694

Signed-off-by: romanetar <roman_ag@hotmail.com>

### CODE REVIEW

File: app/Http/Controllers/Apis/Protected/Summit/Factories/Registration/PromoCodesValidationRulesFactory.php [hunk 1]
```diff
@@ -154,7 +154,9 @@ public static function buildForUpdate(array $payload = []): array
             'valid_until_date'     => 'nullable|required_with:valid_since_date|date_format:U|after:valid_since_date',
             'allowed_ticket_types' => 'sometimes|int_array',
             'badge_features'       => 'sometimes|int_array',
-            'badge_features_apply_to_all_tix_retroactively' => 'sometimes|boolean'
+            'badge_features_apply_to_all_tix_retroactively' => 'sometimes|boolean',
+            'tags'                 => 'sometimes|string_array',
+            'track_id'             => 'nullable|required_with:tags|integer',
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Summit/Factories/Registration/PromoCodesValidationRulesFactory.php:159]: this is not correct
u could add any tag to promo code
tags does not depends on track id
please remove this restriction

File: app/ModelSerializers/Summit/Registration/PromoCodes/SummitRegistrationPromoCodeSerializer.php [hunk 2]
```diff
@@ -69,6 +70,14 @@ public function serialize($expand = null, array $fields = [], array $relations =
             $values['allowed_ticket_types'] = $ticket_types;
         }
 
+        if(in_array('tags', $relations)) {
```
Comment (smarcet) [inline app/ModelSerializers/Summit/Registration/PromoCodes/SummitRegistrationPromoCodeSerializer.php:73]: refactor this serializer to use serialization by reflections as we do for 
https://github.com/OpenStackweb/summit-api/blob/7d55ec41e82c2a40ad771cad25580b2492fb89e3/app/ModelSerializers/AbstractMemberSerializer.php#L91

File: app/Services/Model/Imp/SummitPromoCodeService.php [hunk 3]
```diff
@@ -231,6 +231,23 @@ public function updatePromoCode(Summit $summit, $promo_code_id, array $data, Mem
 
             }
 
+            // tags
+            if (isset($data['tags']) && isset($data['track_id'])) {
```
Comment (smarcet) [inline app/Services/Model/Imp/SummitPromoCodeService.php:235]: please remove this track_id restrictions
promo code should allow any tag

File: app/Services/Model/Imp/SummitPromoCodeService.php [hunk 4]
```diff
@@ -231,6 +231,23 @@ public function updatePromoCode(Summit $summit, $promo_code_id, array $data, Mem
...
+                $track = $summit->getPresentationCategory(intval($data['track_id']));
+                if (is_null($track))
+                    throw new EntityNotFoundException(sprintf("Track with id %s does not belongs to summit id %s.", $data['track_id'], $summit->getId()));
+
+                $promo_code->clearTags();
+
+                foreach ($data['tags'] as $tag_value) {
+                    $tag = $track->getAllowedTagByVal($tag_value);
```
Comment (smarcet) [inline app/Services/Model/Imp/SummitPromoCodeService.php:243]: also if the tag does not exists on tag repository should be created

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review comments

File: app/ModelSerializers/Summit/Registration/PromoCodes/SummitRegistrationPromoCodeSerializer.php (hunk 2 above)
Comment (romanetar) [reply app/ModelSerializers/Summit/Registration/PromoCodes/SummitRegistrationPromoCodeSerializer.php:73]: done

File: app/Http/Controllers/Apis/Protected/Summit/Factories/Registration/PromoCodesValidationRulesFactory.php (hunk 1 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/Factories/Registration/PromoCodesValidationRulesFactory.php:159]: fixed

File: app/Services/Model/Imp/SummitPromoCodeService.php (hunk 4 above)
Comment (romanetar) [reply app/Services/Model/Imp/SummitPromoCodeService.php:243]: done

**Comment by romanetar** [comment]
@smarcet all review comments fixed.

**Review by smarcet — APPROVED** [review]
LGTM

**Pull request merged**